File: 28-03-2024/City With the Smallest Number of Neighbors at a Threshold Distance.cpp

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++
class Solution {
  public:
    int findCity(int n, int m, vector<vector<int>>& edges, int distanceThreshold) {
        // Your Code Start
        vector<vector<pair<int, int>>> adj(n);
        for(int i = 0; i < m; i++){
            adj[edges[i][0]].push_back({edges[i][2], edges[i][1]});
            adj[edges[i][1]].push_back({edges[i][2], edges[i][0]});
        }
        
        int ans = -1, mini = INT_MAX;
        for(int i = 0; i < n; i++){
         
            priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
            
            vector<int> vis(n, 0);
            
            pq.push({0, i});
            int cnt = 0;
            
            while(!pq.empty()){
                
                auto it = pq.top();
                pq.pop();
                
                int dist = it.first;
                int node = it.second;
                
                if(vis[node] == 1) continue;
                
                vis[node] = 1;
                
                if(dist > distanceThreshold) continue;
                
                cnt++;
                
                for(auto nbr_pair : adj[node]){
                    
                    int nbr_dist = nbr_pair.first;
                    int nbr_node = nbr_pair.second;
                    
                    
                    if(vis[nbr_node] == 0){
                        
                        pq.push({dist + nbr_dist, nbr_node});
                    }
                }
                
            }
            
            if(cnt <= mini){
                mini = cnt;
                ans = i;
            }
        }
        
        return ans;
      //Your Code End
    }
};


//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        int n, m;
        cin >> n >> m;
        vector<vector<int>> adj;
        // n--;
        for (int i = 0; i < m; ++i) {
            vector<int> temp;
            for (int j = 0; j < 3; ++j) {
                int x;
                cin >> x;
                temp.push_back(x);
            }
            adj.push_back(temp);
        }

        int dist;
        cin >> dist;
        Solution obj;
        cout << obj.findCity(n, m, adj, dist) << "\n";
    }
}

// } Driver Code Ends

File: 28-03-2024/City With the Smallest Number of Neighbors at a Threshold Distance.java
//{ Driver Code Starts
// Initial Template for Java

import java.util.*;
import java.io.*;
import java.lang.*;

class Main {
    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();

        while (t-- > 0) {

            int n = sc.nextInt();
            int m = sc.nextInt();
            int[][] adj = new int[m][3];

            for (int i = 0; i < m; i++) {

                for (int j = 0; j < 3; j++) {
                    adj[i][j] = sc.nextInt();
                }
            }

            int dist = sc.nextInt();
            Solution obj = new Solution();
            int ans = obj.findCity(n, m, adj, dist);
            System.out.println(ans);
        }
    }
}

// } Driver Code Ends


// User function Template for Java
//Your Code Start

class Solution {
      int findCity(int n, int m, int[][] e,int d)
      {
          int[][] ma=new int[n][n];
          for(int[] r:ma){
          Arrays.fill(r,10001);}
          for(int i=0;i<n;i++){
              ma[i][i]=0;
          }
          for(int[]it:e){
              ma[it[0]][it[1]]=ma[it[1]][it[0]]=it[2];
          }
          for(int k=0;k<n;k++){
              for(int i=0;i<n;i++){
                  for(int j=0;j<n;j++){
                      ma[i][j]=Math.min(ma[i][j],ma[i][k]+ma[k][j]);
                  }
              }
          }
          int ans=-1;
          int ne=n+1;
          for(int i=0;i<n;i++){
              int cnt=0;
              for(int j=0;j<n;j++){
                  if(ma[i][j]<=d)cnt++;
              }
              if(cnt<=ne){
                  ne=cnt;
                  ans=i;
              }
          }
          return ans;
        //Your Code End
      
      }
}
